### PR TITLE
Interrupt FV-optimization based on the gradient norm in image space

### PR DESCRIPTION
**Pull request by thoklei**
In every step of the optimization, we calculate the inverse FFT of the gradient, take its L2 norm and select the maximum over the batch. We then smooth it exponentially and interrupt the optimization if the ratio between the means of two consecutive windows of this sequence falls below a threshold (1.1). This seems to yield colorful, high-contrast images of sufficiently high activation for Googlenet and Resnet50.

### CODE REVIEW

File: lucent/optvis/render.py [hunk 1]
```diff
@@ -81,6 +114,20 @@ def render_vis(
 
     transform_f = transform.compose(transforms)
 
+    # TODO this only works for optimisation in Fourier Space now, support pixel space as well
```
Comment (zimmerrol) [inline lucent/optvis/render.py]: Add a check here that throws an error if a non-fourier parametrization is used.

File: lucent/optvis/render.py [hunk 2]
```diff
@@ -81,6 +114,20 @@ def render_vis(
...
+    def get_mean_px_grad(): 
+        fft_grads = params[0].grad
```
Comment (zimmerrol) [inline lucent/optvis/render.py]: Pass gradient as function argument.

File: lucent/optvis/render.py [hunk 3]
```diff
@@ -130,7 +202,7 @@ def closure():
         show(tensor_to_img_array(image_f()))
     elif show_image:
         view(image_f())
-    return images
+    return images #, smooth_pixelwise_grads
```
Comment (zimmerrol) [inline lucent/optvis/render.py]: Consider adding an argument that allows to return the smoothed gradients as well?

File: lucent/optvis/render.py [hunk 4]
```diff
@@ -81,6 +114,20 @@ def render_vis(
...
 
+    # TODO this only works for optimisation in Fourier Space now, support pixel space as well
+    def get_mean_px_grad(): 
+        fft_grads = params[0].grad
+
+        if type(fft_grads) is not torch.complex64:
+            fft_grads = torch.view_as_complex(fft_grads)
+
+        _, _, h, w = image_shape
+        # NOTE in lucent.param.spatial.fft_image(), they scale the spectrum first
```
Comment (zimmerrol) [inline lucent/optvis/render.py:127]: Wrap this computation in `torch.no_grad` to save memory.

**Review by zimmerrol — CHANGES_REQUESTED** [review]
Thanks for creating this! I had a couple of small comments, otherwise everything looks good!

File: lucent/optvis/render.py (hunk 1 above)
Comment (thoklei) [reply lucent/optvis/render.py]: did it by simply checking the dimensionality of params, which seems unelegant, but don't see a nicer way...

**Review by zimmerrol — APPROVED** [review]
LGTM
